### ISSUE TITLE
Neither bluetooth overlay nor hciattach will work with onboard BT module - aarch kernel build

### ISSUE THREAD

**Issue by gearhead** [moonman/MyPKGBUILDs#12]
I went through the config again and am trying a couple things that *may* cause this, but I am not so sure. 
I had BT working with an older kernel (5.4.45) with the Pi_bluetooth (AUR) package and bluez-utils-compat and something happened and it does not seem to work any more. 
What prompted my invetigation this was this post: [https://github.com/Arkq/bluez-alsa/issues/205](url)
I verified that this does work for armv6 and armv7. Pretty slick when it works. I have re-built the kernel with 4.5.51 and re-installed it and cannot get this to work. For some reason, the old brcm43438 service also fails whicl leads me to believe something in the config may be the culprit. I'll keep digging and will provide a config if I get it working. 
Looking for any pointers if you have them.

**Comment by gearhead** [comment]
A bit more work. I re-built the kernel with the config8 here and get the same result. If I put an armv7 image in this Pi, the onboard bluetooth is on and acts normal. With the aarch64 image, I get no onboard bluetooth. If I try to use the hciattach command from the brcm43438.service from here [https://aur.archlinux.org/pi-bluetooth.git](url), I get this:
`Aug 01 11:30:27 rune64 systemd[1]: Started Broadcom BCM43438 bluetooth HCI.
Aug 01 11:30:27 rune64 kernel: uart-pl011 3f201000.serial: no DMA platform data
Aug 01 11:30:30 rune64 systemd[1]: systemd-timedated.service: Succeeded.
Aug 01 11:30:37 rune64 hciattach[1580]: Initialization timed out.
Aug 01 11:30:37 rune64 hciattach[1580]: bcm43xx_init
Aug 01 11:30:37 rune64 systemd[1]: brcm43438.service: Main process exited, code=exited, status=1/FAILURE
Aug 01 11:30:37 rune64 systemd[1]: brcm43438.service: Failed with result 'exit-code'.`
I have also tried to get the overlay to work as outlined here: [https://github.com/raspberrypi/linux/pull/3682](url) and just get this in dmesg:
`[    2.045285] uart-pl011 3f201000.serial: cts_event_workaround enabled
[    2.045357] 3f201000.serial: ttyAMA0 at MMIO 0x3f201000 (irq = 66, base_baud = 0) is a PL011 rev2
[    2.046767] 3f215040.serial: ttyS0 at MMIO 0x0 (irq = 61, base_baud = 50000000) is a 16550
`
Is there something in the config that needs to be set at build time to get onboard BT to work? It seems like something is missing. When I plug in a bt dongle, it works just fine.

**Comment by gearhead** [comment]
I tried this and still the same result: the module does not activate at boot and also the command :
`hciattach -n /dev/ttyAMA0 bcm43xx 921600 noflow -`
results in this error:
```
Aug 02 10:05:31 rune64 kernel: uart-pl011 3f201000.serial: no DMA platform data
Aug 02 10:05:41 rune64 hciattach[1604]: Initialization timed out.
Aug 02 10:05:41 rune64 hciattach[1604]: bcm43xx_init
Aug 02 10:05:41 rune64 systemd[1]: brcm43438.service: Main process exited, code=exited, status=1/FAILURE
Aug 02 10:05:41 rune64 systemd[1]: brcm43438.service: Failed with result 'exit-code'.
```
this is the diff to the current config8. I noted that these differences to the current Arch ARM Raspberrypi4 config (and a lot of modules that are not built on aarch64 that are built on armv7).
```
54c54
< CONFIG_GENERIC_IRQ_DEBUGFS=y
---
> # CONFIG_GENERIC_IRQ_DEBUGFS is not set
1610c1610
< # CONFIG_BT_HCIBTSDIO is not set
---
> CONFIG_BT_HCIBTSDIO=m
1615c1615
< # CONFIG_BT_HCIUART_BCSP is not set
---
> CONFIG_BT_HCIUART_BCSP=y
1617c1617
< # CONFIG_BT_HCIUART_LL is not set
---
> CONFIG_BT_HCIUART_LL=y
1631c1631
< CONFIG_BT_WILINK=m
---
> # CONFIG_BT_WILINK is not set
1633a1634
> CONFIG_BT_HCIRSI=m
```

**Comment by moonman** [comment]
Thanks. These weren't enabled in defconfig and I didn't check to see if something might be missing assuming defconfig would take care of at least hardware. I'll enable these and rebuild the kernel today.

**Comment by gearhead** [comment]
I just went through it. I grabbed the kernel source for 5.4.51 and did 
` make -C rpi-5.4.51/ ARCH=arm64 O=bo_5.4/ bcmrpi3_defconfig `
copied the generated .config as config8 and rebuilt it. I rebooted and BT now works when enabled from the config.txt. There are a lot of missing modules, though I am not certain they will make much difference for me. I attached the config I used for the kernel build to this reply.

[config_RPiOS.txt](https://github.com/moonman/MyPKGBUILDs/files/5013178/config_RPiOS.txt)

**Comment by moonman** [comment]
are you using it for RPi4 or 3? the defconfig  you use is for RPi3

**Comment by gearhead** [comment]
I am building this for rpi2-4. This is for runeaudio and I have been testing out an aarch64 image. I have a Pi4 as a build machine (with distccd) and have been testing it out on a PiB3+. What is to be gained if I use a defconfig for pi4? I want to be able to have this image boot on pi2 1.2 thru pi4 if possible. I have tried it on B3, B3+ and it seems to work fine. Should I use a defconfig for pi4 even if I want it to boot and run on all the other Pis that can boot aarch64?

(edit1) I think I now figured out what can go wrong. This kernel will not boot on my RPi4 only the 3B+. I just generated a defconfig for the bcm2711_defconfig and am building a new kernel. 
I guess I do not completely understand the defconfig and the 'buildarch' and how that affect things. I have built all my kernels to date for aarch64 with 'buildarch = 8' and a slightly modified version of your build package (I add a patch for a DAC that is not in the mainline) and have been able to boot the image on B3, B3+, B4 without issue. That is my goal, to minimize the number of images. Right now I have armv6, armv7 and aarch64 images that will cover all Pis for me. 

(edit2) made a bcm2711_defconfig and the built image boots on B3 and B4, again. Note that BT and WiFi (iwd) works fine.

**Comment by robsalasco** [comment]
Hi all! @gearhead Can you post your new bcm2711_defconfig please? @moonman will you merge the new changes to enable bluetooth :)?

**Comment by gearhead** [comment]
I enabled a bunch of audio codecs, since this is for RuneAudio, and nlmon, as I am trying to debug an iwd issue., but this does boot on B3+ and B4. I assume it boots on B2 and B3, too, but have not tried. 


This is the Config8 zipped.

[config_RPiOS_B4.zip](https://github.com/moonman/MyPKGBUILDs/files/5042188/config_RPiOS_B4.zip)

**Comment by moonman** [comment]
I got a new kernel built with some added options, see if that solves it for you.

**Comment by robsalasco** [comment]
Works for me without issues :)

**Comment by gearhead** [comment]
Latest update breaks something. I did a "pacman -Syu" and it took me a while to get it back running. For whatever reason, the initramfs is now called kernel8_uboot.img? I created a new 'from scratch' image and noted that change and copied over the files and now it boots and mostly works, again. But now bluetooth does not work, again. No messages in dmesg to any effect about hci... Any ideas? Still running kernel 5.4.51-1. I even made a new initramfs and called it kernel8_uboot.img and still no BT.
edit - I figured it out and had to copy the dtbs from /boot/broadcom to /boot and rebooted and BT was back.
do we need to 'adjust' things so that we can update the system yet still use a RPi kernel for all the functionality? It is strange that this is in config.txt:
`initramfs initramfs-linux.img followkernel`
yet there is no file by that name any more.
